lib/preprocess.js

```diff
@@ -0,0 +1,42 @@
+import assert from 'node:assert'
+import { ethers } from 'ethers'
+
+export const preprocess = async ({
+  rounds,
+  cid,
+  roundIndex,
+  web3Storage,
+  logger
+}) => {
+  const measurements = (await fetchMeasurements(web3Storage, cid))
+    .filter(measurement => {
+      try {
+        assertValidMeasurement(measurement)
+        return true
+      } catch (err) {
+        logger.error('Invalid measurement', err.message, measurement)
+        return false
+      }
+    })
+  logger.log(`Fetched ${measurements.length} valid measurements`)
+
+  if (!rounds[roundIndex]) {
+    rounds[roundIndex] = []
+  }
+  rounds[roundIndex].push(...measurements)
+}
+
+const fetchMeasurements = async (web3Storage, cid) => {
+  const res = await web3Storage.get(cid)
+  const files = await res.files()
+  const measurements = JSON.parse(await files[0].text())
+  return measurements
```

Comment:
Let's discuss attack vectors.

- What happens when the CID points to a file that's not a JSON? Is our entire service going to crash on an unhandled error?
- What if `measurements` is not an array, or the array contain items that are not valid measurement objects?
- What if the first CID file is very large, e.g. several GB of JSON? Will it trigger denial-of-service?

Reply:
> > What happens when the CID points to a file that's not a JSON? Is our entire service going to crash on an unhandled error?
> > What if measurements is not an array, or the array contain items that are not valid measurement objects?
> 
> These will only create a `console.error`, see `index.js`.

I see; this is good.

Can you please report these errors to Sentry as well? (I am ok to do that after this PR is landed if you prefer, but please create a follow-up issue so that we don't forget.)

> > What if the first CID file is very large, e.g. several GB of JSON? Will it trigger denial-of-service?
> 
> This is an attack vector 👍 It can make the service crash (it won't lead to inconsistent data or anything like that). In the meridian design where you run fraud detection on only a single measurement, we could make it so only one measurement ever is stored in memory. We could also use a streaming parser here.
> 
> In the current spark design, evaluation happens in memory, given all measurements of the round. Therefore I think as long as this is the case, we're not adding attack vectors at this part. Eventually we should make the service more stable by probably moving that calculation into postgres, or find some other way to not require looking at all of the data.

Can we reject CIDs where the data is larger than a reasonable threshold? If we expect our measurement service to commit 1000 CIDs at a time, then I think 1MB is a reasonable limit.

Such change does not prevent an attacker from committing many CIDs per round to cause the out-of-memory problem later in our evaluation pipeline, so maybe it's not worth implementing this measure on its own.

Let's open a follow-up GH issue where we can discuss different approaches and agree on what is meaningful to implement now & what we want to implement later.

Reply:
> Can you please report these errors to Sentry as well? (I am ok to do that after this PR is landed if you prefer, but please create a follow-up issue so that we don't forget.)

https://github.com/filecoin-station/spark-evaluate/issues/4

> Can we reject CIDs where the data is larger than a reasonable threshold? If we expect our measurement service to commit 1000 CIDs at a time, then I think 1MB is a reasonable limit.

Such change does not prevent an attacker from committing many CIDs per round to cause the out-of-memory problem later in our evaluation pipeline, so maybe it's not worth implementing this measure on its own.

Let's open a follow-up GH issue where we can discuss different approaches and agree on what is meaningful to implement now & what we want to implement later.

Got it, can you please open that issue, as it's Spark specific?